### ISSUE TITLE
[BUG]: when using `loss_function`, Template Expression cannot have operators outside of `binary_operators`

### ISSUE THREAD

**Issue by Moelf** [MilesCranmer/SymbolicRegression.jl#380]
### What happened?

consider this:
```julia
atlas_template = TemplateStructure{(:b, :e)}(
     ((; b, e), (x,)) -> b(x)^e(x)
)

function test_loss(tree, dataset::Dataset{T,L}, options)::L where {T,L}
  prediction, flag = eval_tree_array(tree, dataset.X, options)
  if !flag
      return L(Inf)
  end
  return sum((prediction .- dataset.y).^2)
end

sr_model2 = SRRegressor(
    niterations=150,
    binary_operators=[+, -, *],
    expression_type=TemplateExpression,
    expression_options=(; structure=atlas_template),
    loss_function = test_loss
)

sr_mach2 = machine(sr_model2, X, ys)

fit!(sr_mach2)
```

right now it errors:

### Version

v1.2.0

### Operating System

Linux

### Interface

Jupyter Notebook

### Relevant log output

```shell
[ Info: Training machine(SRRegressor(defaults = nothing, …), …).
┌ Error: Problem fitting the machine machine(SRRegressor(defaults = nothing, …), …). 
└ @ MLJBase ~/.julia/packages/MLJBase/7nGJF/src/machines.jl:694
[ Info: Running type checks... 
[ Info: Type checks okay. 

Operator ^ not found in operators for expression type ComposableExpression{Float64, DynamicExpressions.NodeModule.Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*)}, Tuple{}}, variable_names::Nothing}} with binary operators (+, -, *)

Stacktrace:
  [1] apply_operator(op::typeof(^), l::ComposableExpression{Float64, DynamicExpressions.NodeModule.Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*)}, Tuple{}}, variable_names::Nothing}}, r::ComposableExpression{Float64, DynamicExpressions.NodeModule.Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*)}, Tuple{}}, variable_names::Nothing}})
    @ DynamicExpressions.ExpressionAlgebraModule ~/.julia/packages/DynamicExpressions/fxs7F/src/ExpressionAlgebra.jl:85
  [2] ^(l::ComposableExpression{Float64, DynamicExpressions.NodeModule.Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*)}, Tuple{}}, variable_names::Nothing}}, r::ComposableExpression{Float64, DynamicExpressions.NodeModule.Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*)}, Tuple{}}, variable_names::Nothing}})
    @ DynamicExpressions.ExpressionAlgebraModule ~/.julia/packages/DynamicExpressions/fxs7F/src/ExpressionAlgebra.jl:119
  [3] (::var"#15#16")(::@NamedTuple{b::ComposableExpression{Float64, DynamicExpressions.NodeModule.Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*)}, Tuple{}}, variable_names::Nothing}}, e::ComposableExpression{Float64, DynamicExpressions.NodeModule.Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*)}, Tuple{}}, variable_names::Nothing}}}, ::Vector{ComposableExpression{Float32, DynamicExpressions.NodeModule.Node{Float32}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*)}, Tuple{}}, variable_names::Nothing}}})
    @ Main ./In[28]:2
```

### Extra Info

removing custom `loss_function` fixs this

**Comment by MilesCranmer** [comment]
Sorry, `loss_function` and `TemplateExpression` aren't compatible at the moment. This should raise a better error though.

**Comment by MilesCranmer** [comment]
Actually I guess they are compatible. But it requires your template structure to only use operators that are found in the operator set, otherwise `get_tree` won't be able to assemble the `TemplateExpression` into a single binary tree.

But a better solution would simply be to allow for `loss_function` to take the `TemplateExpression` directly.

**Comment by MilesCranmer** [comment]
In principle it's easy to implement this. I just have no idea what to do for the API. Should it be a different parameter, like `loss_function_expression`? Or should there be some flag you set that says the full expression should be passed, rather than the tree? Or something else? Feels like there's not really a clean way to add this at the moment.

**Comment by Moelf** [comment]
I guess I didn't understand why loss function needs to be expressed in terms of SR expressions, if user doesn't need to take derivative, can we not view loss function as a blackbox and just evaluate it?

**Comment by MilesCranmer** [comment]
Yeah it can definitely be a black box. The error you saw is just when it was trying to extract a single `AbstractExpressionNode` from the template expression, to pass to the loss.

The difference is whether the user needs to define it in terms of `AbstractExpressionNode` input (current API), or `AbstractExpression` (proposed API). The latter would let you skip this issue. I just don’t know the right API for declaring one way or the other.

**Comment by Moelf** [comment]
I though I have found a hack workaround:
```julia
    binary_operators=[+, -, *, ^],
    unary_operators=[log],
    constraints = [(^)=>(0, 0), log=>0],
```

But unfortunately it still shows up in the HoF

**Comment by MilesCranmer** [comment]
Actually I think you can do `complexity_of_operators=[(^) => 100]`, try that?

But yeah I should still fix this at the source.

**Comment by Moelf** [comment]
that somehow doesn't quite work -- it results in bad equations being kept in HoF at the end

For example, before, it was finding formulas like:

![Image](https://github.com/user-attachments/assets/316af069-cc5a-4ac9-82e3-fb2623d391ab)


Now it finds things like (and it's barely using the variables for some reason):

![Image](https://github.com/user-attachments/assets/dfb6dce5-f760-4d2e-91b8-11d7c6bfa057)

**Comment by MilesCranmer** [comment]
Maybe it likes using `log` and `^` as stepping stones towards other expressions (this is a real thing that can happen - its soft regularization vs hard constraint).

Maybe try `complexity_of_operators=[(^) => 20, log => 20]`? Then it can still sometimes use them in intermediate expressions.

**Comment by MilesCranmer** [comment]
P.S., be sure to check out the tuning page here: https://ai.damtp.cam.ac.uk/pysr/tuning/. Some people find using `adaptive_parsimony_scaling=20.0` can also help too.

**Comment by Moelf** [comment]
I actually tried it with => 10, suspecting the same thing as you pointed out. When setting the complexity to 5, it's not enough to eliminate them. I guess this route would involve a bit more tuning we can afford (time-wise).

**Comment by MilesCranmer** [comment]
Another workaround would be to use the template expression as the loss, rather than using the separate loss function parameter. So, e.g., set `y=0`, and then pass the actual `y` as a feature of `X`, and return the squared residual from the template expression.
